### PR TITLE
`iree` submodule

### PR DESCRIPTION
**Pull request by makslevental**
This PR checks in `iree` as a submodule instead of using `sync_deps.py`. I don't know what the original theory of `sync_deps.py` but this simplifies dev and deploy.

### CODE REVIEW

**Review by jtuyls — COMMENTED** [review]
LGTM, but I also don't know why we went for the script in the first place. Maybe @Abhishek-Varma knows?

**Comment by Abhishek-Varma** [comment]
> LGTM, but I also don't know why we went for the script in the first place. Maybe @Abhishek-Varma knows?

I don't know but can only speculate based on the doc comment :-
```
 Casual developers and CI bots invoke this to do the most efficient checkout of dependencies.
```
Perhaps @stellaraccident can chime in here.

**Comment by makslevental** [comment]
>  Casual developers and CI bots invoke this to do the most efficient checkout of dependencies.

Sure but I think no one was/is taking advantage of that (I added skipping torch-mlir but only last week). For that, a simpler way is for iree to make torch-mlir opt-in https://github.com/iree-org/iree/issues/18409 .

**Review by newling — APPROVED** [review]
I don't know what this reason was for doing it with `sync_deps.py` either, but I can't see why not to do it this way with submodules if it works and has as much flexibility in terms of clone performance. Nice one!

**Comment by makslevental** [comment]
@newling @jtuyls @Abhishek-Varma can I get a re-review. I changed the clone commands/instructions in order to save people time/space.

File: README.md [hunk 1]
```diff
@@ -14,24 +14,46 @@ interfacing the AMD AIE accelerator to IREE.
...
+  -c submodule."third_party/stablehlo".update=none \
+  -c submodule."src/runtime_src/core/common/aiebu".update=none \
+  clone \
+  --recursive \
+  --depth 1 \
+  --shallow-submodules \
+  https://github.com/nod-ai/iree-amd-aie.git
 ```
 
+which has the effect of not cloning entire repo histories *and* skipping nested submodules that we currently do not need.
```
Comment (newling) [inline README.md:46]:
```suggestion
which has the effect of not cloning entire repo histories and skipping nested submodules that we currently do not need.
```

File: README.md [hunk 2]
```diff
@@ -14,24 +14,46 @@ interfacing the AMD AIE accelerator to IREE.
 
 ## Developer Setup
 
-These instructions assume that you have an appropriate IREE checkout side by side
-with this repository have an IREE build setup in an `iree-build` directory that
-is also a sibling. This is not a requirement, but instructions will need to be
-changed for different paths.
+**Strong recommendation**: check the CI scripts @ [.github/workflows](.github/workflows) - they do a fresh/clean 
+checkout and build and are exercised on every commit *and* are written such that they're simple enough to be read 
+by even a non-CI expert.
```
Comment (newling) [inline README.md:19]:
```suggestion
by a non-CI expert.
```

**Review by jtuyls — COMMENTED** [review]
I was looking at the log and I think the `--depth 1` flag is not working:

```
Run git \
  git \
    -c submodule."third_party/torch-mlir".update=none \
    -c submodule."third_party/stablehlo".update=none \
    -c submodule."src/runtime_src/core/common/aiebu".update=none \
    clone \
    --recursive \
    --shallow-submodules \
    --depth 1 \
    -b $BRANCH_NAME $REPO_ADDRESS .
  shell: /usr/bin/bash -e {0}
  env:
    CACHE_DIR: /_work/iree-amd-aie/iree-amd-aie/.container-cache
    CACHE_KEY: linux-build-test-cpp-asserts-manylinux-v[2](https://github.com/nod-ai/iree-amd-aie/actions/runs/10671927949/job/29578539280?pr=733#step:4:2)-733/merge-405
    BRANCH_NAME: makslevental/iree-submodule
    REPO_ADDRESS: https://github.com/nod-ai/iree-amd-aie
Cloning into '.'...
Submodule 'third_party/XRT' (https://github.com/nod-ai/XRT.git) registered for path 'third_party/XRT'
Submodule 'third_party/aie-rt' (https://github.com/Xilinx/aie-rt.git) registered for path 'third_party/aie-rt'
Submodule 'third_party/bootgen' (https://github.com/Xilinx/bootgen.git) registered for path 'third_party/bootgen'
Submodule 'third_party/iree' (https://github.com/iree-org/iree.git) registered for path 'third_party/iree'
Submodule 'third_party/mlir-air' (https://github.com/nod-ai/mlir-air.git) registered for path 'third_party/mlir-air'
Submodule 'third_party/openssl' (https://github.com/viaduck/openssl-cmake.git) registered for path 'third_party/openssl'
Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/XRT'...
Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/aie-rt'...
Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/bootgen'...
Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/iree'...
Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/mlir-air'...
warning: --depth is ignored in local clones; use file:// instead.
done.
```

**Review by newling — APPROVED** [review]
Looks good, thanks for updating the README.

**Comment by makslevental** [comment]
> I was looking at the log and I think the `--depth 1` flag is not working:
> 
> ```
> Run git \
>   git \
>     -c submodule."third_party/torch-mlir".update=none \
>     -c submodule."third_party/stablehlo".update=none \
>     -c submodule."src/runtime_src/core/common/aiebu".update=none \
>     clone \
>     --recursive \
>     --shallow-submodules \
>     --depth 1 \
>     -b $BRANCH_NAME $REPO_ADDRESS .
>   shell: /usr/bin/bash -e {0}
>   env:
>     CACHE_DIR: /_work/iree-amd-aie/iree-amd-aie/.container-cache
>     CACHE_KEY: linux-build-test-cpp-asserts-manylinux-v[2](https://github.com/nod-ai/iree-amd-aie/actions/runs/10671927949/job/29578539280?pr=733#step:4:2)-733/merge-405
>     BRANCH_NAME: makslevental/iree-submodule
>     REPO_ADDRESS: https://github.com/nod-ai/iree-amd-aie
> Cloning into '.'...
> Submodule 'third_party/XRT' (https://github.com/nod-ai/XRT.git) registered for path 'third_party/XRT'
> Submodule 'third_party/aie-rt' (https://github.com/Xilinx/aie-rt.git) registered for path 'third_party/aie-rt'
> Submodule 'third_party/bootgen' (https://github.com/Xilinx/bootgen.git) registered for path 'third_party/bootgen'
> Submodule 'third_party/iree' (https://github.com/iree-org/iree.git) registered for path 'third_party/iree'
> Submodule 'third_party/mlir-air' (https://github.com/nod-ai/mlir-air.git) registered for path 'third_party/mlir-air'
> Submodule 'third_party/openssl' (https://github.com/viaduck/openssl-cmake.git) registered for path 'third_party/openssl'
> Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/XRT'...
> Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/aie-rt'...
> Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/bootgen'...
> Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/iree'...
> Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/mlir-air'...
> warning: --depth is ignored in local clones; use file:// instead.
> done.
> ```
> 
> (https://github.com/nod-ai/iree-amd-aie/actions/runs/10671927949/job/29578539280?pr=733)

Ya I saw that and that's why I added -shallow-submodules. I'm not 100% on whether this works correctly but 🤷‍♂️

File: README.md [hunk 3]
```diff
@@ -14,24 +14,46 @@ interfacing the AMD AIE accelerator to IREE.
...
 ## Building (along with IREE)
 
 ### Just show me the CMake
 
+From the checkout of the repo:
+
 ```
-cmake -B $WHERE_YOU_WOULD_LIKE_TO_BUILD -S $IREE_REPO_SRC_DIR \
--DIREE_CMAKE_PLUGIN_PATHS=$IREE_AMD_AIE_REPO_SRC_DIR -DIREE_BUILD_PYTHON_BINDINGS=ON \
+cmake -B $WHERE_YOU_WOULD_LIKE_TO_BUILD -S third_party/iree \
```
Comment (jtuyls) [inline README.md:55]:
```suggestion
cd iree-amd-aie
cmake -B $WHERE_YOU_WOULD_LIKE_TO_BUILD -S third_party/iree \
```

**Review by jtuyls — APPROVED** [review]
LGTM

**Comment by jtuyls** [comment]
> > I was looking at the log and I think the `--depth 1` flag is not working:
> > ```
> > Run git \
> >   git \
> >     -c submodule."third_party/torch-mlir".update=none \
> >     -c submodule."third_party/stablehlo".update=none \
> >     -c submodule."src/runtime_src/core/common/aiebu".update=none \
> >     clone \
> >     --recursive \
> >     --shallow-submodules \
> >     --depth 1 \
> >     -b $BRANCH_NAME $REPO_ADDRESS .
> >   shell: /usr/bin/bash -e {0}
> >   env:
> >     CACHE_DIR: /_work/iree-amd-aie/iree-amd-aie/.container-cache
> >     CACHE_KEY: linux-build-test-cpp-asserts-manylinux-v[2](https://github.com/nod-ai/iree-amd-aie/actions/runs/10671927949/job/29578539280?pr=733#step:4:2)-733/merge-405
> >     BRANCH_NAME: makslevental/iree-submodule
> >     REPO_ADDRESS: https://github.com/nod-ai/iree-amd-aie
> > Cloning into '.'...
> > Submodule 'third_party/XRT' (https://github.com/nod-ai/XRT.git) registered for path 'third_party/XRT'
> > Submodule 'third_party/aie-rt' (https://github.com/Xilinx/aie-rt.git) registered for path 'third_party/aie-rt'
> > Submodule 'third_party/bootgen' (https://github.com/Xilinx/bootgen.git) registered for path 'third_party/bootgen'
> > Submodule 'third_party/iree' (https://github.com/iree-org/iree.git) registered for path 'third_party/iree'
> > Submodule 'third_party/mlir-air' (https://github.com/nod-ai/mlir-air.git) registered for path 'third_party/mlir-air'
> > Submodule 'third_party/openssl' (https://github.com/viaduck/openssl-cmake.git) registered for path 'third_party/openssl'
> > Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/XRT'...
> > Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/aie-rt'...
> > Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/bootgen'...
> > Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/iree'...
> > Cloning into '/_work/iree-amd-aie/iree-amd-aie/third_party/mlir-air'...
> > warning: --depth is ignored in local clones; use file:// instead.
> > done.
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > (https://github.com/nod-ai/iree-amd-aie/actions/runs/10671927949/job/29578539280?pr=733)
> 
> Ya I saw that and that's why I added -shallow-submodules. I'm not 100% on whether this works correctly but 🤷‍♂️

Yeah, not sure, it says to add `file://...`, maybe something to try at some point, but it looks like the clone goes reasonably fast.. at least faster than `sync_deps.py`!

**Comment by makslevental** [comment]
> Yeah, not sure, it says to add `file://...`, maybe something to try at some point, but it looks like the clone goes reasonably fast.. at least faster than `sync_deps.py`!

Ya the thing is I have no clue where to add this... Anyway I'll put a TODO there and make an issue

File: README.md (hunk 2 above)
Comment (makslevental) [reply README.md:19]: Bruh lol no allowance for any tone

**Pull request merged**